staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -0,0 +1,192 @@
+type Validator struct {
+	Items      *Validator
+	Properties map[string]Validator
+
+	AdditionalProperties *Validator
+
+	compiledRules []CompilationResult
+
+	// Program compilation is pre-checked at CRD creation/update time, so we don't expect compilation to fail
+	// they are recompiled and added to this type, and it does, it is an internal bug.
+	// But if somehow we get any compilation errors, we track them and then surface them as validation errors.
+	compilationErr error
+
+	// isResourceRoot is true if this validator node is for the root of a resource. Either the root of the
+	// custom resource being validated, or the root of an XEmbeddedResource object.
+	isResourceRoot bool
+}
+
+// NewValidator returns compiles all the CEL programs defined in x-kubernetes-validations extensions
+// of the Structural schema and returns a custom resource validator.
+func NewValidator(s *schema.Structural) *Validator {
+	return validator(s, true)
+}
+
+func validator(s *schema.Structural, isResourceRoot bool) *Validator {
+	compiledRules, err := Compile(s, isResourceRoot)
+	if s.Items != nil {
+		if i := validator(s.Items, s.Items.XEmbeddedResource); i != nil {
+			return &Validator{compiledRules: compiledRules, compilationErr: err, isResourceRoot: isResourceRoot, Items: i}
+		}
+	}
+	if len(s.Properties) > 0 {
+		properties := make(map[string]Validator, len(s.Properties))
+		for k, prop := range s.Properties {
+			if p := validator(&prop, prop.XEmbeddedResource); p != nil {
+				properties[k] = *p
+			}
+		}
+		if len(properties) > 0 {
+			return &Validator{compiledRules: compiledRules, compilationErr: err, isResourceRoot: isResourceRoot, Properties: properties}
+		}
+	}
+	if s.AdditionalProperties != nil && s.AdditionalProperties.Structural != nil {
+		if ap := validator(s.AdditionalProperties.Structural, s.AdditionalProperties.Structural.XEmbeddedResource); ap != nil {
+			return &Validator{compiledRules: compiledRules, compilationErr: err, isResourceRoot: isResourceRoot, AdditionalProperties: ap}
+		}
+	}
+	if len(compiledRules) > 0 {
+		return &Validator{compiledRules: compiledRules, compilationErr: err, isResourceRoot: isResourceRoot}
+	}
+
+	return nil
+}
+
+// Validate validates all x-kubernetes-validations rules in Validator against obj and returns any errors.
+func (s *Validator) Validate(fldPath *field.Path, sts *schema.Structural, obj interface{}) field.ErrorList {
+	if s == nil || obj == nil {
+		return nil
+	}
+
+	errs := s.validateExpressions(fldPath, sts, obj)
+	switch obj := obj.(type) {
+	case []interface{}:
+		return append(errs, s.validateArray(fldPath, sts, obj)...)
+	case map[string]interface{}:
+		return append(errs, s.validateMap(fldPath, sts, obj)...)
+	}
+	return errs
+}
+
+func (s *Validator) validateExpressions(fldPath *field.Path, sts *schema.Structural, obj interface{}) (errs field.ErrorList) {
+	if s.isResourceRoot {
+		sts = model.WithTypeAndObjectMeta(sts)
+	}
+	activation := NewValidationActivation(obj, sts)
```

Comment:
move these after the compilationErr check so we don't do this work if we're just going to return an error

Reply:
also return early `if len(s.compiledRules) == 0` (there might not be expressions at this level, so avoid augmenting `sts` and constructing `activation` unless we'll actually use them)